lib/EarlyFunctionAnalysis/EarlyFunctionAnalysis.cpp

```diff
@@ -1973,54 +2034,30 @@ bool EarlyFunctionAnalysis::runOnModule(Module &M) {
   using CFEPA = CFEPAnalyzer<FunctionAnalysisResults>;
   CFEPA Analyzer(M, &GCBI, ABIRegisters, Properties, Binary);
 
-  // Interprocedural analysis over the collected functions in post-order
-  // traversal (leafs first).
-  while (!CFEPQueue.empty()) {
-    BasicBlockNode *EntryNode = CFEPQueue.pop();
-    revng_log(EarlyFunctionAnalysisLog,
-              "Analyzing Entry: " << EntryNode->BB->getName());
-
-    // Intraprocedural analysis
-    FunctionSummary AnalysisResult = Analyzer.analyze(EntryNode->BB);
-    bool Changed = Properties.registerFunction(getBasicBlockPC(EntryNode->BB),
-                                               std::move(AnalysisResult));
-
-    // If we got improved results for a function, we need to recompute its
-    // callers, and if a caller turns out to be fake, the callers of the fake
-    // function too.
-    if (Changed) {
-      UniquedQueue<BasicBlockNode *> FakeFunctionWorklist;
-      FakeFunctionWorklist.insert(EntryNode);
+  // If requested, recover the control-flow graph and serialize it to the LLVM
+  // IR, without executing the whole analysis.
+  if (!ShouldAnalyzeABI) {
+    Analyzer.recoverCFG();
+  } else {
+    // Interprocedural analysis over the collected functions in post-order
+    // traversal (leafs first).
+    runInterproceduralAnalysis(CFEPQueue, RootNode, Analyzer, Properties);
 
-      while (!FakeFunctionWorklist.empty()) {
-        BasicBlockNode *Node = FakeFunctionWorklist.pop();
-        for (auto *Caller : Node->predecessors()) {
-          if (Caller == RootNode)
-            break;
+    // Propagate results between call-sites and functions
+    interproceduralPropagation(Properties, Binary.Functions);
 
-          if (!Properties.isFakeFunction(getBasicBlockPC(Caller->BB)))
-            CFEPQueue.insert(Caller);
-          else
-            FakeFunctionWorklist.insert(Caller);
-        }
-      }
-    }
+    // Finalize model
+    finalizeModel(GCBI, ABIRegisters, Properties, Binary);
   }
 
+  // Serialize function metadata, including the CFG, to the entry basic blocks
+  // in `root`.
+  serializeFunctionMetadata(GCBI, Properties, Binary);
```

Comment:
I feel like `runInterproceduralAnalysis`, `interproceduralPropagation`, `finalizeModel` and `serializeFunctionMetadata` could become `Analyzer`'s methods and have no arguments at all.

Reply:
[commit [tmp](https://github.com/revng/revng/pull/125/commits/09ada2570036f6e8f3097118fe573320d8392e82)]: I had mixed feelings about making `finalizeModel` (formerly, `commitToModel`) an `Analyzer`'s method, since it is not completely related to the analysis per se. Nonetheless, given that it was using several common fields, I moved `finalizeModel`, together with `runInterproceduralAnalysis` and `interproceduralPropagation` as methods of `Analyzer`. Conversely, I felt more like that `serializeFunctionMetadata` should not be part of them (not related to the analysis at all). Also, I took the chance to simplify and refactor `EarlyFunctionAnalysis::runOnModule` and create a static function `createApproximateCallGraph` that returns the entry-points queue to be analyzed (decided not to bring it into `Analyzer` though, since it sounds more as a pre-phase to be done before instantiating the `Analyzer` object). What do you think?

Reply:
Removed the last part, since the lifetime of different objects is compromised if I move the creation of the call graph out in a static function. Need to think about this.